File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-14.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public TreeNode LowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root.val == p.val || root.val == q.val) return root;
        if (p.val > q.val){
            TreeNode temp = q;
            q = p;
            p = temp;
        }

        if (p.val < root.val){
            if (q.val < root.val){
                return LowestCommonAncestor(root.left, p, q);
            }

            return root;
        }

        return LowestCommonAncestor(root.right, p, q);  
    }
}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-14.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private:
    int dfs(TreeNode* root, int maxVal){
        if (!root) return 0;
        int count = 0;
        if (root->val >= maxVal){
            maxVal = root->val;
            count++;
        }

        return count + dfs(root->left, maxVal) + dfs(root->right, maxVal);

    }
public:
    int goodNodes(TreeNode* root) {
        return dfs(root, INT_MIN);
    }
};
